qualtran/bloqs/rotations/hamming_weight_phasing.py

```diff
@@ -0,0 +1,87 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from functools import cached_property
+from typing import Dict, Set, TYPE_CHECKING
+
+import attrs
+
+from qualtran import GateWithRegisters, Signature
+from qualtran.bloqs.arithmetic import HammingWeightCompute
+from qualtran.bloqs.basic_gates import ZPowGate
+
+if TYPE_CHECKING:
+    from qualtran import BloqBuilder, SoquetT
+    from qualtran.resource_counting.bloq_counts import BloqCountT
+
+
+@attrs.frozen
+class HammingWeightPhasing(GateWithRegisters):
+    r"""Applies $Z^{\theta}$ to every qubit of input register of size `bitsize`.
```

Comment:
theta == exponent below right? I think just pick one (I like Z^\theta) better.

Reply:
Let's pick `exponent` because `ZPowGate` also expects an `exponent` parameter, similar to Cirq